### ISSUE TITLE
Importing twisted.protocols.dns blocks for /dev/random

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11237]
|<img alt="ben's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ben reported|
|-|-|
|Trac ID|trac#856|
|Type|defect|
|Created|2005-01-07 13:54:55Z|

<details><summary>Searchable metadata</summary>

```
trac-id__856 856
type__defect defect
reporter__ben ben
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1105106095000000 1105106095000000
changetime__1112285262000000 1112285262000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__tekNico cc__ben
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ben's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ben commented|
|-|-|

```
#!html
<pre>
The module "twisted.protocols.dns" is imported by "twisted.internet.utils" and 
possibly other modules.

In lines 46-56 the following top level code gets executed unconditionally, 
provided that the module "Crypto.Util.randpool" can not be imported:

    for randSource in ('random', 'srandom', 'urandom'):
        try:
            f = file('/dev/' + randSource)
            f.read(2) # BLOCKS HERE IN THIS BUG
            f.close()
        except:
            pass
        else:
            def randomSource(r = file('/dev/' + randSource, 'rb').read):
                return struct.unpack('H', r(2))[0]
            break

If there is no entropy in /dev/random, the program importing any module that 
eventually imports twisted.protocols.dns will (somewhat mysteriously) block, 
even if it never uses "dns".

On "FreeBSD 4.9-RELEASE" and some Linux distributions I've tested, the rate at 
which the "/dev/random" entropy pool is filled is very slow, particularly on 
servers not used interactively. The end result is the application hangs on 
start until you press Control-C. The program then ignores the break exception 
and continues to run.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I suggest sticking to urandom only (what's "srandom"?) I see no reason why the
extra grade entropy from /dev/random should be necessary in  most cases.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
by r12982
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="teknico's avatar" src="https://avatars.githubusercontent.com/u/172521?s=50" width="50" height="50">](https://github.com/teknico)|@teknico commented|
|-|-|

```
#!html
<pre>
I've just been bitten by this, as ben describes below, except that I was indeed
using twisted.names, and the process using it became very slow, without using
significant CPU.

It's worth noting that this issue has been fixed in 2.0 by removing usage of
/dev/random and /dev/srandom, but those still using 1.3.0 should beware. Guess
it's time to begin testing the fresher beast. :-)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
